### PR TITLE
Rule-based detector and script for gpt segmentation

### PR DESCRIPTION
**Pull request by voidf**
- 加了个容错率更高的compare_breaks_v2，方便人工标注（以及直接编辑GPT标注来快速做手标数据）
- 规则型分段器，因为实现代码过于冗长换了一个文件装
- 一堆围绕gpt的请求以及后处理相关的代码

### CODE REVIEW

File: alignment/rule_based_detector.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            if CHINESE_UNIT_DICT[digit] < unit:
+                unit = CHINESE_UNIT_DICT[digit]
+                num += unit
+            else:
+                unit = CHINESE_UNIT_DICT[digit]
+        elif digit in CHINESE_NUM_DICT:
+            num += CHINESE_NUM_DICT[digit] * unit
+    return num
+
+def read_int(s: str) -> int:
```
Comment (liyongsea) [inline alignment/rule_based_detector.py:35]: you can do return int(s)

File: alignment/rule_based_detector.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    """从s的开头开始读一段连续的数字"""
+    x = 0
+    for c in s:
+        if c.isdigit():
+            x = x * 10 + int(c)
+        else:
+            return x
+    return x
+
+def read_back_int(s: str) -> int:
```
Comment (liyongsea) [inline alignment/rule_based_detector.py:45]: `return int(s.split('.')[-1])`

File: alignment/rule_based_detector.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        if (not nextline) or (not prevline): # 当两行中一行是空行，则拼接
+            return INF
+        if RuleBasedDetector.match_lineno_seg(nextline): # 避免和lineno规则冲突
+            return -INF
+        return 0
+    
+    def detect(self, lines: list[str], **kwargs) -> list[bool]:
+        breaks = [True] * (len(lines) - 1)
+        match_infos = [] # 存(int数字列表号, int文件行号) 这样的二元组
+        line_marker = [] # 可以去掉换行的行下标
```
Comment (liyongsea) [inline alignment/rule_based_detector.py:154]: you can use the same vocabulary - soft_linebreak_indice for line_marker

File: alignment/rule_based_detector.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    for i in s:
+        o = ord(i) - ord(begin_char)
+        if 0 <= o <= 25:
+            return o
+    return -2
+
+
+
+class RuleBasedDetector(HardLineBreakDetector):
+    LINENO_SEG_TOKENS = [
```
Comment (liyongsea) [inline alignment/rule_based_detector.py:80]: lineno是什么意思

File: alignment/rule_based_detector.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    @staticmethod
+    def score_special(prevline: str, nextline: str) -> int:
+        INF = 998244353
+        if (not nextline) or (not prevline): # 当两行中一行是空行，则拼接
+            return INF
+        if RuleBasedDetector.match_lineno_seg(nextline): # 避免和lineno规则冲突
+            return -INF
+        return 0
+    
+    def detect(self, lines: list[str], **kwargs) -> list[bool]:
```
Comment (liyongsea) [inline alignment/rule_based_detector.py:151]: doc string to explain detect

Comment (liyongsea) [reply alignment/rule_based_detector.py:151]: Un example
```
def detect(self, lines: list[str], **kwargs) -> list[bool]:
    """
    Detects the positions of hard line breaks in a list of lines based on rule-based heuristics.

    Args:
        lines (list[str]): A list of strings representing the lines of text.
        **kwargs: Additional keyword arguments (not used in this method).

    Returns:
        list[bool]: A list of boolean values indicating the positions of hard line breaks.
                    True indicates a hard line break should be present at the corresponding index,
                    while False indicates that the line break should be removed (lines should be merged).

    Note:
        The detection process is performed using rule-based heuristics that evaluate various factors,
        such as line numbering patterns, punctuation, line lengths, and linguistic analysis with the help of NLTK.
        The method assigns scores to each line break, and based on the scores, determines whether to keep or remove
        the line break. The line breaks are represented as boolean values in the returned list, where True indicates
        a hard line break and False indicates the line break should be removed.
    """

You could write it in chinese and explain there are two steps. Please make sure you use some key words such as soft line break, hard line break, numbering patterns ..
```

File: alignment/rule_based_detector.py (hunk 1 above)
Comment (voidf) [reply alignment/rule_based_detector.py]: `read_int` is designed for reading a int from a string robustly (so we can use it under multiple cases and combine the rule with the unordered list rules), or we may use more complex logic for LINENO_SEG_TOKENS match.

**Pull request closed without merging**